File: julia/src/YAMLStar.jl
module YAMLStar

import JSON

include("libyamlstar.jl")

mutable struct Runtime
    isolate::Ref{Ptr{Cvoid}}
    thread::Ref{Ptr{Cvoid}}
    error
    closed::Bool

    function Runtime()
        ys = new(Ref{Ptr{Cvoid}}(), Ref{Ptr{Cvoid}}(), nothing, false)
        ys.isolate[] = C_NULL
        ys.thread[] = C_NULL
        libyamlstar.create_isolate(ys.isolate, ys.thread)
        return ys
    end
end

function close(ys::Runtime)
    if !ys.closed
        libyamlstar.tear_down_isolate(ys.thread[])
        ys.closed = true
    end
    return nothing
end

function load(ys::Runtime, input::AbstractString)
    call_yaml(ys, libyamlstar.yamlstar_load, input)
end

function load_all(ys::Runtime, input::AbstractString)
    call_yaml(ys, libyamlstar.yamlstar_load_all, input)
end

function dump(ys::Runtime, value)
    call_json(ys, libyamlstar.yamlstar_dump, value)
end

function dump_all(ys::Runtime, values)
    call_json(ys, libyamlstar.yamlstar_dump_all, values)
end

function version(ys::Runtime)
    unsafe_string(libyamlstar.yamlstar_version(ys.thread[]))
end

function call_yaml(ys::Runtime, func::Function, input::AbstractString)
    handle_response(ys, unsafe_string(func(ys.thread[], String(input))))
end

function call_json(ys::Runtime, func::Function, value)
    handle_response(ys, unsafe_string(func(ys.thread[], JSON.json(value))))
end

function handle_response(ys::Runtime, json_src::AbstractString)
    resp = JSON.parse(json_src)
    ys.error = get(resp, "error", nothing)
    if ys.error !== nothing
        error("libyamlstar: " * get(ys.error, "cause", ""))
    end
    if haskey(resp, "data")
        return resp["data"]
    else
        error("Unexpected response from 'libyamlstar'")
    end
end

end

File: ruby/lib/yamlstar.rb
# frozen_string_literal: true

require "fiddle"
require "fiddle/import"
require "json"

require_relative "yamlstar/version"

class YAMLStar
  Error = Class.new(StandardError)

  LIBYAMLSTAR_VERSION = VERSION

  module LibYAMLStar
    extend Fiddle::Importer

    def self.extension
      case RUBY_PLATFORM
      when /darwin/
        "dylib"
      when /linux/
        "so"
      when /mswin|mingw|cygwin/
        "dll"
      else
        raise Error, "Unsupported platform #{RUBY_PLATFORM} for yamlstar."
      end
    end

    def self.library_names
      base = "libyamlstar.#{extension}"
      [base, "#{base}.#{LIBYAMLSTAR_VERSION}"]
    end

    def self.library_paths
      paths = []
      dev_path = File.expand_path("../../libyamlstar/lib", __dir__)
      paths << dev_path
      env_var = extension == "dll" ? "PATH" : "LD_LIBRARY_PATH"
      paths.concat(ENV.fetch(env_var, "").split(File::PATH_SEPARATOR))
      paths << "/usr/local/lib" unless extension == "dll"
      paths << File.join(ENV.fetch("HOME", ""), ".local", "lib")
      paths.reject(&:empty?)
    end

    def self.find_library
      library_paths.each do |dir|
        library_names.each do |name|
          path = File.join(dir, name)
          return path if File.exist?(path)
        end
      end

      raise Error, <<~ERROR
        Shared library file `libyamlstar.#{extension}` not found
        Search paths: #{library_paths.join(File::PATH_SEPARATOR)}
        Try: curl -sSL https://yamlstar.org/install | LIB=1 bash
      ERROR
    end

    dlload find_library

    extern "int graal_create_isolate(void* params, void** isolate, void** thread)"
    extern "int graal_tear_down_isolate(void* thread)"
    extern "char* yamlstar_load(void* thread, char* yaml)"
    extern "char* yamlstar_load_all(void* thread, char* yaml)"
    extern "char* yamlstar_dump(void* thread, char* json)"
    extern "char* yamlstar_dump_all(void* thread, char* json)"
    extern "char* yamlstar_version(void* thread)"
  end

  def self.load(yaml)
    new.load(yaml)
  end

  def self.load_all(yaml)
    new.load_all(yaml)
  end

  def self.dump(value)
    new.dump(value)
  end

  def self.dump_all(values)
    new.dump_all(values)
  end

  attr_reader :error

  def initialize
    @isolate = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
    thread = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
    rc = LibYAMLStar.graal_create_isolate(nil, @isolate.ref, thread.ref)
    raise Error, "Failed to create GraalVM isolate" unless rc.zero?

    @thread = thread
    @closed = false
  end

  def load(yaml)
    call_yaml(:yamlstar_load, yaml)
  end

  def load_all(yaml)
    call_yaml(:yamlstar_load_all, yaml)
  end

  def dump(value)
    call_json(:yamlstar_dump, value)
  end

  def dump_all(values)
    call_json(:yamlstar_dump_all, values)
  end

  def version
    LibYAMLStar.yamlstar_version(@thread).to_s
  end

  def close
    return if @closed

    rc = LibYAMLStar.graal_tear_down_isolate(@thread)
    raise Error, "Failed to tear down GraalVM isolate" unless rc.zero?

    @closed = true
  end

  private

  def call_yaml(function, input)
    handle_response(LibYAMLStar.public_send(function, @thread, input.to_s).to_s)
  end

  def call_json(function, value)
    handle_response(LibYAMLStar.public_send(function, @thread, JSON.generate(value)).to_s)
  end

  def handle_response(json)
    resp = JSON.parse(json)
    @error = resp["error"]
    raise Error, "libyamlstar: #{@error["cause"]}" if @error
    raise Error, "Unexpected response from 'libyamlstar'" unless resp.key?("data")

    resp["data"]
  end
end
